From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Wed, 5 Feb 2025 01:24:07 +0000
Subject: rockchip: rk3506: Enable use of SPL_OPTEE_IMAGE

Vendor OP-TEE blob provides PSCI that is used to enable CPU cores on
RK3506. The OP-TEE blob is expected to be loaded and executed from
0x1000 and uses memory in the 0-1 MiB area of DRAM.

Add initial support for loading OP-TEE firmware on RK3506. Reserve the
first 1 MiB of DRAM for OP-TEE and use ROCKCHIP_SPL_RESERVE_IRAM to add
padding to the start of SPL text area to allow loading OP-TEE to 0x1000.

Signed-off-by: Austin Lane <vidplace7@gmail.com>
---
 arch/arm/dts/rk3506-u-boot.dtsi | 11 ++++++++++
 arch/arm/dts/rk3506.dtsi        |  8 +++++++
 arch/arm/mach-rockchip/Kconfig  |  4 ++++
 arch/arm/mach-rockchip/sdram.c  |  5 ++++-
 include/configs/rk3506_common.h |  1 +
 5 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/arch/arm/dts/rk3506-u-boot.dtsi b/arch/arm/dts/rk3506-u-boot.dtsi
index 111111111111..222222222222 100644
--- a/arch/arm/dts/rk3506-u-boot.dtsi
+++ b/arch/arm/dts/rk3506-u-boot.dtsi
@@ -29,6 +29,17 @@
 	bootph-all;
 };
 
+#ifdef CONFIG_SPL_OPTEE_IMAGE
+&fit_template {
+	images {
+		op-tee {
+			load = <0x1000>;
+			entry = <0x1000>;
+		};
+	};
+};
+#endif
+
 &grf_pmu {
 	bootph-all;
 };
diff --git a/arch/arm/dts/rk3506.dtsi b/arch/arm/dts/rk3506.dtsi
index 111111111111..222222222222 100644
--- a/arch/arm/dts/rk3506.dtsi
+++ b/arch/arm/dts/rk3506.dtsi
@@ -43,6 +43,7 @@
 			reg = <0xf00>;
 			clocks = <&cru ARMCLK>;
 			device_type = "cpu";
+			enable-method = "psci";
 		};
 
 		cpu1: cpu@f01 {
@@ -50,6 +51,7 @@
 			reg = <0xf01>;
 			clocks = <&cru ARMCLK>;
 			device_type = "cpu";
+			enable-method = "psci";
 		};
 
 		cpu2: cpu@f02 {
@@ -57,6 +59,7 @@
 			reg = <0xf02>;
 			clocks = <&cru ARMCLK>;
 			device_type = "cpu";
+			enable-method = "psci";
 		};
 	};
 
@@ -133,6 +136,11 @@
 		};
 	};
 
+	psci {
+		compatible = "arm,psci-1.0", "arm,psci-0.2";
+		method = "smc";
+	};
+
 	timer {
 		compatible = "arm,armv7-timer";
 		interrupts = <GIC_PPI 13 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>,
diff --git a/arch/arm/mach-rockchip/Kconfig b/arch/arm/mach-rockchip/Kconfig
index 111111111111..222222222222 100644
--- a/arch/arm/mach-rockchip/Kconfig
+++ b/arch/arm/mach-rockchip/Kconfig
@@ -332,6 +332,7 @@ config ROCKCHIP_RK3506
 	select BOARD_LATE_INIT
 	select DM_REGULATOR_FIXED
 	select DM_RESET
+	imply ARM_SMCCC if SPL_OPTEE_IMAGE
 	imply BOOTSTD_FULL
 	imply DM_RNG
 	imply ENV_RELOC_GD_ENV_ADDR
@@ -352,6 +353,7 @@ config ROCKCHIP_RK3506
 	imply SPL_FIT_SIGNATURE
 	imply SPL_LOAD_FIT
 	imply SPL_OF_CONTROL
+	imply SPL_OPTEE_IMAGE if SPL_FIT
 	imply SPL_PINCTRL
 	imply SPL_RAM
 	imply SPL_REGMAP
@@ -359,6 +361,7 @@ config ROCKCHIP_RK3506
 	imply SPL_SYSCON
 	imply SYS_ARCH_TIMER
 	imply SYSRESET
+	imply SYSRESET_PSCI if SPL_OPTEE_IMAGE
 	help
 	  The Rockchip RK3506 is a ARM-based SoC with a tri-core Cortex-A7.
 
@@ -679,6 +682,7 @@ config ROCKCHIP_STIMER_BASE
 
 config ROCKCHIP_SPL_RESERVE_IRAM
 	hex "Size of IRAM reserved in SPL"
+	default 0x40000 if ROCKCHIP_RK3506 && SPL_OPTEE_IMAGE
 	default 0x0
 	depends on SPL
 	help
diff --git a/arch/arm/mach-rockchip/sdram.c b/arch/arm/mach-rockchip/sdram.c
index 111111111111..222222222222 100644
--- a/arch/arm/mach-rockchip/sdram.c
+++ b/arch/arm/mach-rockchip/sdram.c
@@ -320,7 +320,7 @@ int dram_init_banksize(void)
 		gd->bd->bi_dram[0].size = ram_top - gd->bd->bi_dram[0].start;
 	}
 #else
-#ifdef CONFIG_SPL_OPTEE_IMAGE
+#if defined(CONFIG_SPL_OPTEE_IMAGE) && !defined(CFG_SYS_SDRAM_RESERVE)
 	struct tos_parameter_t *tos_parameter;
 
 	tos_parameter = (struct tos_parameter_t *)(CFG_SYS_SDRAM_BASE +
@@ -343,6 +343,9 @@ int dram_init_banksize(void)
 	}
 #else
 	gd->bd->bi_dram[0].start = CFG_SYS_SDRAM_BASE;
+#ifdef CFG_SYS_SDRAM_RESERVE
+	gd->bd->bi_dram[0].start += CFG_SYS_SDRAM_RESERVE;
+#endif
 	gd->bd->bi_dram[0].size = top - gd->bd->bi_dram[0].start;
 #endif
 #endif
diff --git a/include/configs/rk3506_common.h b/include/configs/rk3506_common.h
index 111111111111..222222222222 100644
--- a/include/configs/rk3506_common.h
+++ b/include/configs/rk3506_common.h
@@ -11,6 +11,7 @@
 #define CFG_IRAM_BASE			0xfff80000
 
 #define CFG_SYS_SDRAM_BASE		0
+#define CFG_SYS_SDRAM_RESERVE		SZ_1M
 #define SDRAM_MAX_SIZE			0xc0000000
 
 #ifndef ROCKCHIP_DEVICE_SETTINGS
-- 
Armbian

